### PR TITLE
Root when

### PR DESCRIPTION
**Pull request by gigabo**
The `<RootElement listen={...}>` is really concise with gated emitters, but for general stores it's easier to have the gate and the emitter separate.  It's also nice to be able to have two gates even with gated emitters.

So, this adds `<RootElement when={...}>`.  The promise can resolve with a props object, but it doesn't _need_ to.  Render-time props are the union of `listen` emitter props and `when` promise props, with the promise winning collisions.

### CODE REVIEW

**Comment by gigabo** [comment]
This also fleshes out the test pages packages a little bit.  Might make the goal there a little clearer.

File: packages/react-server/core/components/RootElement.js [hunk 1]
```diff
@@ -147,3 +148,26 @@ RootElement.scheduleRender = function(element) {
...
+	// It may be a gated emitter, but we want to make sure we squeeze
+	// props out of it from the beginning if it's not.
+	// Finally gate on the `when`.
+	return Q(listen && RootElement.installListener(element, listen))
+		.then(el => el && (rendered = el))
+		.then(() => when)
+		.then(childProps => childProps
+			?React.cloneElement(rendered, {childProps})
+			:rendered
+		)
```
Comment (roblg) [inline packages/react-server/core/components/RootElement.js]: I'm unsure of the complete flow here, but it seems like if `listen` isn't specified on props, `rendered` will never be set, which seems like it means the promise would resolve w/ `undefined`?

Comment (gigabo) [inline packages/react-server/core/components/RootElement.js]: `rendered` is initialized to `element` upon declaration.

Comment (roblg) [inline packages/react-server/core/components/RootElement.js]: :facepalm:

**Comment by roblg** [comment]
LGTM. Not that that matters that much, because I'm obviously incapable of reading today.

**Comment by gigabo** [comment]
Thanks for reviewing.  I've been working on a change set based on this, so I'll be glad to get it merged. :)

**Comment by gigabo** [comment]
BTW - I'd also like to start using https://github.com/redfin/react-server/pull/78, so please have a look there if you get a chance.

**Pull request merged**